[Feature] Include "Auto-sklearn metrics over time curve" as function in auto-sklearn

OliEfr opened the pull request:
#1128 

Done:
- [x] Add performance_over_time_ property to automl-based objects (just like cv_results). This attribute contains a dataframe, which can be used for plotting easily.
- [x] Adjusted [this example](https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_pandas_train_test.html#sphx-glr-examples-40-advanced-example-pandas-train-test-py) to show, how this property can be used for plotting.
- [x] Made some further changes to the docs  regarding the new property
- [x] Add pytest to requirements.txt
- [x] successfully run pre-commit locally  

ToDo:
- [x] write tests and run tests
- [x] build the docs

franchuterivera commented on requirements.txt:
```diff
@@ -2,6 +2,7 @@ setuptools
 
 numpy>=1.9.0
 scipy>=0.14.1
+pytest
```
I think it will be preferred not to have this here. The reason for this is that [this line](https://github.com/automl/auto-sklearn/blob/275d0d6b20d16822252d8b50bf71b1c787187f09/setup.py#L29) already provides this functionality.

This functionality can be accessed if you do something like `pip install -e .[test]` for example.

franchuterivera commented on autosklearn/automl.py:
```diff
@@ -1343,6 +1343,66 @@ def score(self, X, y):
+            })
+            #append test-scores, if data for test_loss are available. This is the case, if X_test and y_test where provided.
+            if 'test_loss' in run_value.additional_info:
+                test_score = metric._optimum - (metric._sign * run_value.additional_info['test_loss'])
```
Probably we can be robust against not having `test_loss` in the additional_info dictionary.

What I mean, is that we can do something like:
```
test_loss = np.nan
if test_loss' in run_value.additional_info:
    est_loss = <further code here>
```

Because, `X_test` might not be passed to Auto-sklearn

OliEfr replied:
I suppose you mean 

```
test_score = np.nan
if 'test_loss' in run_value.additional_info:
    test_score = <further code here>
```

Otherwise I dont know what you mean. My implementation is already robust against not having the key `'test_loss'`, which is the case when `X_test` is not passed. That is why I am checking for `'test_loss' in run_value.additional_info`

franchuterivera commented on examples/40_advanced/example_pandas_train_test.py:
```diff
@@ -1,26 +1,24 @@
-*auto-sklearn* can automatically encode categorical columns using a label/ordinal encoder.
-This example highlights how to properly set the dtype in a DataFrame for this to happen,
-and showcase how to input also testing data to autosklearn.
 The X_train/y_train arguments to the fit function will be used to fit the scikit-learn model,
```
Auto-sklearn fits both scikit-learn models and an ensemble made from those models. 

I think it makes sense to leave the previous message so that the users have this in mind?

OliEfr commented:
I added tests and fixed a bug.

My changes in the docs build successfully - however I get some memory allocation errors when building examples in docs. I think that might be due to my setup.

Let me know what you thinkg @franchuterivera

franchuterivera commented on examples/40_advanced/example_pandas_train_test.py:
```diff
@@ -1,26 +1,29 @@
-It also shows the training and test performance of the ensemble build using the best
-performing models (under ensemble_optimization_score and ensemble_test_score respectively).
-
-There is also support to manually indicate the feature types (whether a column is categorical
```
Can you please also preserve this message, that gives an intuition about the feature types?

OliEfr commented:
@franchuterivera  done

franchuterivera commented on test/test_automl/test_estimators.py:
```diff
@@ -257,6 +258,42 @@ def test_type_of_target(mock_estimator):
                     "binary targets")
 
 
+def test_performance_over_time(tmp_dir, output_dir):
```
This test looks great, but there is one thing I will like to suggest, which is try to re-use the existing test.

The point of my comment is that unit testing takes a big amount of time, 30+ mins, where the biggest contribution comes from fitting an AutoSklearnClassifier.

What if we instead of creating a new `test_performance_over_time` we add your checks `include_single_scores` and `includes_all_scores` into some existing functions. Maybe one suggestion will be:

* `includes_all_scores` --> [after this line](https://github.com/automl/auto-sklearn/blob/2e58f9df6eb39c3439bef40b390710b7504884c4/test/test_automl/test_estimators.py#L507)
* `include_single_scores` --> [after this line](https://github.com/automl/auto-sklearn/blob/2e58f9df6eb39c3439bef40b390710b7504884c4/test/test_automl/test_estimators.py#L546)

Also, maybe it makes sense to check if the dataframe is not empty and the embedded scores make sense (make sense means that the performance is increasing where we get a cumulative score or greater than 0.5. Something else that you think makes sense)?

franchuterivera commented:
Thanks for the changes! Unit testing is running in here through github actions. If we see a failure, I can help you further with it in case you need help.

OliEfr commented:
@franchuterivera My intention was to have all tests at one place - just as with `test_cv_results` - but yeah I understand the objections.

I have moved everything to already existing tests, apart from `include_single_scores`, because this requires `ensemble_size=0` and I didnt want to but it, where `cv_results_` is not tested.

In addition, I wrote a test for checking the plausibility of the dataframe.  However, I found it difficult to address

> and the embedded scores make sense (make sense means that the performance is increasing where we get a cumulative score or greater than 0.5.

cause I am not sure, whether thats necessarily the case (or I misunderstand your proposal). The only score, which should be increasing all the time is `ensemble_optimization_score`. Also, I don't feel comfortable setting a fixed performance treshold.

franchuterivera commented on examples/40_advanced/example_pandas_train_test.py:
```diff
@@ -1,21 +1,29 @@
 ===============================
-Test and Train data with Pandas
+Performance-over-time plot
 ===============================
```
```suggestion
==========================
```

franchuterivera commented on examples/40_advanced/example_pandas_train_test.py:
```diff
@@ -1,21 +1,29 @@
 # -*- encoding: utf-8 -*-
 """
 ===============================
```
```suggestion
==========================
```

franchuterivera commented:
I think everything looks good, only some changes to the header of the example. What I mean is that for the HTML rendering we expect the same number of `=` as characters in the title.
For example it used to be:
```
===============================
Test and Train data with Pandas
```

It now has to be:
```
==========================
Performance-over-time plot
```

OliEfr commented:
Okay, that was only a minor fix. I did that.

eddiebergman commented:
Hi @OliEfr, apologies for the delayed response, this has been on the radar for some time.

I've done a review of the code and it seems okay and something we want to merge, the only thing I would have to ask you to do is rebase to the most current development branch as it's been updated a few times since. I don't imagine there will be many/any conflicts but if there are, I am happy to help you through it if you need.

To prevent the PR having its 'file changes' messed up as can happen during rebasing, you can either create a new PR and just tag this one or follow this [guide](https://akrabat.com/the-beginners-guide-to-rebasing-your-pr/). I've created an issue (#1205) which will detail these steps in the future as we want to make contributing as friction-less as possible in the future :)

OliEfr commented:
@eddiebergman thanks for getting back. I will have a look later!
Oli

OliEfr commented:
Hey,

okay I performed a rebase on development-branch.

However, I created a new branch on my end (NewFeature). This way I didnt have to force push to my master.

I created a new PR #1219 . Please check this one out!

Oli

eddiebergman commented:
Closing as this is continued in the rebase version at #1219
